lib/shopify_api/rest/base.rb

```diff
@@ -374,22 +393,34 @@ def save(update_object: false)
 
       sig { returns(T::Hash[String, String]) }
       def attributes_to_update
-        original_state_for_update = original_state.reject do |attribute, _|
+        updatable_attributes = original_state.reject do |attribute, _|
           self.class.read_only_attributes&.include?("@#{attribute}".to_sym)
         end
 
-        diff = HashDiff::Comparison.new(
-          deep_stringify_keys(original_state_for_update),
-          deep_stringify_keys(to_hash(true)),
-        ).left_diff
+        stringified_updatable_attributes = deep_stringify_keys(updatable_attributes)
+        stringified_new_attributes = deep_stringify_keys(to_hash(true))
+        ShopifyAPI::Utils::AttributesComparator.compare(
+          stringified_updatable_attributes,
+          stringified_new_attributes,
+        )
+      end
 
-        diff.each do |attribute, value|
-          if value.is_a?(Hash) && value[0] == HashDiff::NO_VALUE
-            diff[attribute] = send(attribute)
-          end
+      sig { params(http_method: Symbol).returns(T::Hash[String, T.untyped]) }
+      def build_required_attributes(http_method:)
+        required_attributes = {}
+
+        primary_key_value = send(self.class.primary_key)
+        unless primary_key_value.nil?
+          required_attributes[self.class.primary_key] = primary_key_value
         end
 
-        diff
+        path_ids = deduce_path_ids(http_method)
+        path_ids&.each do |path_id|
+          path_id_value = send(path_id)
+          required_attributes[path_id.to_s] = path_id_value unless path_id_value.nil?
+        end
+
+        required_attributes
```

Comment:
this makes the assumption that the primary key and anything required path params might also be required in the body payload